### PR TITLE
Add "Copy branch name" button to pull request page

### PR DESCRIPTION
**Pull request by saitho**
As a developer I often check out branches from a pull request, needing to select and copy the branch name.
This pull request adds a copy button next to the head branch name to make that easier.

copy-icon-new

### CODE REVIEW

File: templates/repo/issue/view_title.tmpl
Comment (wxiaoguang) [inline templates/repo/issue/view_title.tmpl]: Some questions here:
1. What if there is `.HeadBranchHTMLURL` and `headHref ` was already converted to HTML?
2. There is a trailing space: `">%s</a> "` 
3. We do not need to introduce `cursor-pointer`, framework already provides:
```
a.ui.label {
  cursor: pointer;
}
```

File: modules/markup/sanitizer.go
Comment (wxiaoguang) [inline modules/markup/sanitizer.go]: Comment and code doesn't match.

File: modules/markup/sanitizer.go
Comment (wxiaoguang) [inline modules/markup/sanitizer.go]: Maybe we can use another func/filter (I just forgot the name) instead of `Str2html` to mark the string is safe already, then we do not need to change the sanitizer. Otherwise there are many edge cases for SVG and the code can be very complex for all cases.

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [reply templates/repo/issue/view_title.tmpl]: It probably shouldn't be a label thought, just a SVG icon, in which case the method used here is fine.

**Comment by silverwind** [comment]
Any screenshots? I'm worried about vertical alignment on the icon.

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [reply templates/repo/issue/view_title.tmpl]: 1. That's fine. The icon is just appended to whatever is in $headHref. In my test case, `.HeadBranchHTMLURL` was set.
2. That's intentional. Otherwise there would be no or very litte space between the icon and the text "into".
3. It was a `<span>` before I changed it into an `<a>`, which is why I added a new class. But I'm also not sure about it being a label.

File: modules/markup/sanitizer.go
Comment (saitho) [reply modules/markup/sanitizer.go]: Oh yeah. Later I noticed there already was a clipboard functionality, so I dropped by own data attribute `data-copy-text`. Will change that comment.

File: modules/markup/sanitizer.go
Comment (saitho) [reply modules/markup/sanitizer.go]: I spend an hour figuring out why the icon is not rendered via `printf`. ^^
I think I already tried using `Safe`, but I'll try to look into it again.

Comment (silverwind) [reply modules/markup/sanitizer.go]: Yeah these new rules seem too specific and unnecessary, I think it shouldn't even go through the sanitizer in this case, e.g. avoid `printf` usage and use normal template syntax.

Comment (saitho) [reply modules/markup/sanitizer.go]: I don't think I can avoid `printf` as branch name and the icon need to be used in a translation.

Comment (saitho) [reply modules/markup/sanitizer.go]: Got it.

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [reply templates/repo/issue/view_title.tmpl]: 3) I needed to adjust the vertical position of the icon a bit, so I moved the cursor styling there as well. The additional space was also removed as I set a small margin-right.

**Review by wxiaoguang — APPROVED** [review]

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [inline templates/repo/issue/view_title.tmpl]:
```suggestion
		{{$headHref = printf "%s <a data-clipboard-text=\"%s\">%s</a>" $headHref .HeadTarget (svg "octicon-copy" 14)}}
```

File: web_src/less/_repository.less
Comment (silverwind) [inline web_src/less/_repository.less]:
```suggestion
          vertical-align: middle;
          position: relative;
          top: -2px;
          margin-right: 1px;
```

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [inline templates/repo/issue/view_title.tmpl:46]: Why this change?

**Comment by silverwind** [comment]
Would really like to see some UI feedback when it has copied, I guess a bit of JS would be needed to show a "Copied!" tooltip.

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [reply templates/repo/issue/view_title.tmpl:46]: Otherwise the icon is removed by HTML sanitization.

**Comment by silverwind** [comment]
Here's a patch that adds interactivity:

````patch
diff --git a/options/locale/locale_en-US.ini b/options/locale/locale_en-US.ini
index 2cb1bcd1a..decc2b0ef 100644
--- a/options/locale/locale_en-US.ini
+++ b/options/locale/locale_en-US.ini
@@ -923,11 +923,13 @@ fork_from_self = You cannot fork a repository you own.
 fork_guest_user = Sign in to fork this repository.
 watch_guest_user = Sign in to watch this repository.
 star_guest_user = Sign in to star this repository.
 copy_link = Copy
-copy_link_success = Link has been copied
+copy_link_success = Link has been copied!
 copy_link_error = Use ⌘C or Ctrl-C to copy
-copied = Copied OK
+copy = Copy
+copy_success = Copied!
+copy_error = Copy failed
 unwatch = Unwatch
 watch = Watch
 unstar = Unstar
 star = Star
diff --git a/templates/repo/issue/view_title.tmpl b/templates/repo/issue/view_title.tmpl
index 6495325e2..54d48854a 100644
--- a/templates/repo/issue/view_title.tmpl
+++ b/templates/repo/issue/view_title.tmpl
@@ -33,9 +33,9 @@
                {{$headHref := .HeadTarget|Escape}}
                {{if .HeadBranchHTMLURL}}
                        {{$headHref = printf "<a href=\"%s\">%s</a>" .HeadBranchHTMLURL $headHref}}
                {{end}}
-               {{$headHref = printf "%s <a data-clipboard-text=\"%s\">%s</a>" $headHref .HeadTarget (svg "octicon-copy")}}
+               {{$headHref = printf "%s <a data-clipboard-text=\"%s\" data-content=\"%s\" data-original=\"%s\" data-success=\"%s\" data-error=\"%s\">%s</a>" $headHref .HeadTarget ($.i18n.Tr "repo.copy") ($.i18n.Tr "repo.copy") ($.i18n.Tr "repo.copy_success") ($.i18n.Tr "repo.copy_error") (svg "octicon-copy" 14)}}
                {{$baseHref := .BaseTarget|Escape}}
                {{if .BaseBranchHTMLURL}}
                        {{$baseHref = printf "<a href=\"%s\">%s</a>" .BaseBranchHTMLURL $baseHref}}
                {{end}}
diff --git a/web_src/js/index.js b/web_src/js/index.js
index 74aca6913..726461420 100644
--- a/web_src/js/index.js
+++ b/web_src/js/index.js
@@ -2673,8 +2673,9 @@ $(document).ready(async () => {
   $('.ui.progress').progress({
     showActivity: false
   });
   $('.poping.up').popup();
+  $('[data-clipboard-text][data-content]').popup();
   $('.top.menu .poping.up').popup({
     onShow() {
       if ($('.top.menu .menu.transition').hasClass('visible')) {
         return false;
````

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [reply templates/repo/issue/view_title.tmpl:46]: Ok, seems odd that we even have to go through sanitation for this but I guess if it works, it's fine.

**Comment by silverwind** [comment]
> 
> 
> @silverwind We do not need `$('[data-clipboard-text][data-content]').popup();` in `index.js`, `clipboard.js` already did it.

It's necessary to have the popup show on initial click. I tried adding `class="poping up"` but that ran into sanitizer issues again, that's why I had added it there.

**Comment by silverwind** [comment]
Patch updated to use only one `.popup()` call.

File: web_src/less/_repository.less
Comment (saitho) [reply web_src/less/_repository.less]: Applied

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [reply templates/repo/issue/view_title.tmpl]: Applied

File: web_src/js/features/clipboard.js
Comment (wxiaoguang) [inline web_src/js/features/clipboard.js]: Why we need this change? if there is no `data-original`, `btn.dataset.original` should be `undefined`, will it cause problems?

Comment (saitho) [reply web_src/js/features/clipboard.js]: I don't like the idea of setting a string property to a undefined or uncontrolled value. ^^

Comment (wxiaoguang) [reply web_src/js/features/clipboard.js]: If it doesn't cause any problem, please leave it as it is. This file has been refactored in another branch, we do not need to introduce unnecessary changes.

Comment (silverwind) [reply web_src/js/features/clipboard.js]: This code needs to be refactored anyways to get rid of `data-original`. If the value comes out as `undefined`, that value will be visible in the popup which is rather ugly.

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [inline templates/repo/issue/view_title.tmpl]:
```suggestion
		{{$headHref = printf "%s <a class=\"poping up\" data-success=\"%s\" data-error=\"%s\" data-clipboard-text=\"%s\">%s</a>" $headHref (.i18n.Tr "repo.copy_branch_success") (.i18n.Tr "repo.copy_branch_error") .HeadTarget (svg "octicon-copy" 14)}}
```

**Comment by silverwind** [comment]
Actually, I figured it out, I was just missing the `\` to escape the `"`, it's now working with the `poping up` class.

File: web_src/js/features/clipboard.js
Comment (wxiaoguang) [reply web_src/js/features/clipboard.js]: Agree, we should get rid of `data-original`, we can do it in this PR, or leave it to another.

There is only one tag uses `data-original=xxx`

The refactor in my mind:

```
  const contentOriginal = btn.dataset.content;
  $(btn).popup('destroy');
  btn.dataset.content = btn.dataset.success;
  $(btn).popup('show');
  btn.dataset.content = contentOriginal;  // question again: doesn't the `popup` support `undefined` value? if no support, we should use null or delete the key instead.
```

**Comment by silverwind** [comment]
One more patch that should fix the remaining issues:

````patch
diff --git a/options/locale/locale_en-US.ini b/options/locale/locale_en-US.ini
index d374493df..22fe9ae2f 100644
--- a/options/locale/locale_en-US.ini
+++ b/options/locale/locale_en-US.ini
@@ -926,8 +926,9 @@ watch_guest_user = Sign in to watch this repository.
 star_guest_user = Sign in to star this repository.
 copy_link = Copy
 copy_link_success = Link has been copied
 copy_link_error = Use ⌘C or Ctrl-C to copy
+copy_branch = Copy
 copy_branch_success = Branch name has been copied
 copy_branch_error = Use ⌘C or Ctrl-C to copy
 copied = Copied OK
 unwatch = Unwatch
diff --git a/templates/repo/clone_buttons.tmpl b/templates/repo/clone_buttons.tmpl
index 9eb7fa781..0a86e586f 100644
--- a/templates/repo/clone_buttons.tmpl
+++ b/templates/repo/clone_buttons.tmpl
@@ -13,9 +13,9 @@
 {{else if and (not .DisableSSH) (or $.IsSigned $.ExposeAnonSSH)}}
        <input id="repo-clone-url" value="{{if $.PageIsWiki}}{{$.WikiCloneLink.SSH}}{{else}}{{$.CloneLink.SSH}}{{end}}" readonly>
 {{end}}
 {{if or (not $.DisableHTTP) (and (not $.DisableSSH) (or $.IsSigned $.ExposeAnonSSH))}}
-       <button class="ui basic icon button poping up" id="clipboard-btn" data-original="{{.i18n.Tr "repo.copy_link"}}" data-success="{{.i18n.Tr "repo.copy_link_success"}}" data-error="{{.i18n.Tr "repo.copy_link_error"}}" data-content="{{.i18n.Tr "repo.copy_link"}}" data-variation="inverted tiny" data-clipboard-target="#repo-clone-url">
+       <button class="ui basic icon button poping up" id="clipboard-btn" data-success="{{.i18n.Tr "repo.copy_link_success"}}" data-error="{{.i18n.Tr "repo.copy_link_error"}}" data-content="{{.i18n.Tr "repo.copy_link"}}" data-variation="inverted tiny" data-clipboard-target="#repo-clone-url">
                {{svg "octicon-paste"}}
        </button>
 {{end}}
 {{if not (and $.DisableHTTP $.DisableSSH)}}
diff --git a/templates/repo/issue/view_title.tmpl b/templates/repo/issue/view_title.tmpl
index 8b5bcfb50..839a4638f 100644
--- a/templates/repo/issue/view_title.tmpl
+++ b/templates/repo/issue/view_title.tmpl
@@ -33,9 +33,9 @@
                {{$headHref := .HeadTarget|Escape}}
                {{if .HeadBranchHTMLURL}}
                        {{$headHref = printf "<a href=\"%s\">%s</a>" .HeadBranchHTMLURL $headHref}}
                {{end}}
-               {{$headHref = printf "%s <a data-success=\"%s\" data-error=\"%s\" data-clipboard-text=\"%s\">%s</a>" $headHref (.i18n.Tr "repo.copy_branch_success") (.i18n.Tr "repo.copy_branch_error") .HeadTarget (svg "octicon-copy" 14)}}
+               {{$headHref = printf "%s <a class=\"poping up\" data-content=\"%s\" data-success=\"%s\" data-error=\"%s\" data-clipboard-text=\"%s\">%s</a>" $headHref (.i18n.Tr "repo.copy_branch") (.i18n.Tr "repo.copy_branch_success") (.i18n.Tr "repo.copy_branch_error") .HeadTarget (svg "octicon-copy" 14)}}
                {{$baseHref := .BaseTarget|Escape}}
                {{if .BaseBranchHTMLURL}}
                        {{$baseHref = printf "<a href=\"%s\">%s</a>" .BaseBranchHTMLURL $baseHref}}
                {{end}}
diff --git a/web_src/js/features/clipboard.js b/web_src/js/features/clipboard.js
index 737a8caf3..af5626439 100644
--- a/web_src/js/features/clipboard.js
+++ b/web_src/js/features/clipboard.js
@@ -1,18 +1,22 @@
 const selector = '[data-clipboard-target], [data-clipboard-text]';

 // TODO: replace these with toast-style notifications
 function onSuccess(btn) {
+  if (!btn.dataset.content) return;
   $(btn).popup('destroy');
+  const oldContent = btn.dataset.content;
   btn.dataset.content = btn.dataset.success;
   $(btn).popup('show');
-  btn.dataset.content = btn.dataset.original ?? '';
+  btn.dataset.content = oldContent;
 }
 function onError(btn) {
+  if (!btn.dataset.content) return;
+  const oldContent = btn.dataset.content;
   $(btn).popup('destroy');
   btn.dataset.content = btn.dataset.error;
   $(btn).popup('show');
-  btn.dataset.content = btn.dataset.original ?? '';
+  btn.dataset.content = oldContent;
 }

 export default async function initClipboard() {
   for (const btn of document.querySelectorAll(selector) || []) {
````

**Comment by wxiaoguang** [comment]
The `data-original="{{.i18n.Tr "repo.copy_link"}}"` in `clone_buttons.tmpl` should also be removed, no code uses it anymore.

File: templates/repo/issue/view_title.tmpl
Comment (zeripath) [inline templates/repo/issue/view_title.tmpl]: .HeadTarget can contain a ". 

You must protect against this.

**Review by zeripath — CHANGES_REQUESTED** [review]

File: templates/repo/issue/view_title.tmpl
Comment (wxiaoguang) [reply templates/repo/issue/view_title.tmpl]: Oh yes, it should be `(.HeadTarget | Escape)`

File: templates/repo/issue/view_title.tmpl
Comment (silverwind) [inline templates/repo/issue/view_title.tmpl]:
```suggestion
		{{$headHref = printf "%s <a class=\"poping up\" data-content=\"%s\" data-success=\"%s\" data-error=\"%s\" data-clipboard-text=\"%s\" data-variation=\"inverted tiny\">%s</a>" $headHref (.i18n.Tr "repo.copy_branch") (.i18n.Tr "repo.copy_branch_success") (.i18n.Tr "repo.copy_branch_error") (.HeadTarget | Escape) (svg "octicon-copy" 14)}}
```

make it the same style as repo url copy:

image

**Review by silverwind — APPROVED** [review]
LGTM after above suggestion is merged.

File: templates/repo/issue/view_title.tmpl
Comment (zeripath) [inline templates/repo/issue/view_title.tmpl]: This is probably going to need to become:

```suggestion
			{{$baseHref = printf "<a href=\"%s\">%s</a>" (.BaseBranchHTMLURL | Escape) $baseHref}}
```

File: templates/repo/issue/view_title.tmpl
Comment (zeripath) [inline templates/repo/issue/view_title.tmpl]: This is probably going to need to become

```suggestion
			{{$headHref = printf "<a href=\"%s\">%s</a>" (.HeadBranchHTMLURL | Escape) $headHref}}
```

Comment (wxiaoguang) [reply templates/repo/issue/view_title.tmpl]: `HeadBranchHTMLURL` comes from  `GetHeadBranchHTMLURL returns the HTML URL of the head branch`.

The name `XxxHTMLURL` shows that it is already a HTML safe URL I think.

Comment (zeripath) [reply templates/repo/issue/view_title.tmpl]: Just because it's safe in terms of being url-encoded doesn't mean that it's safe to go in an attribute. & will not being escaped in this.

Comment (wxiaoguang) [reply templates/repo/issue/view_title.tmpl]: Well, that's true, the name of XxxHTMLURL is pretty misleading.

Comment (wxiaoguang) [reply templates/repo/issue/view_title.tmpl]: Just read the code, XxxHTMLURL only returns a valid URL, nothing with HTML .... Maybe we should rename these functions to clear up the ambiguity.

**Review by zeripath — APPROVED** [review]

**Pull request merged**
